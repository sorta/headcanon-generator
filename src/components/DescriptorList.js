import React, { Component } from 'react';
import PropTypes from 'prop-types';

class DescriptorList extends Component {
  static propTypes = {
    descriptors: PropTypes.object,
    deleteDescriptor: PropTypes.func,
  };
  static defaultProps = {
    descriptors: {},
    deleteDescriptor: () => {},
  };

  render() {
    return (
      <ul className="DescriptorList">
        {Object.keys(this.props.descriptors).map(key => (
          <li key={key} className="DescriptorList-item">
            <span className="DescriptorList-item-text">{this.props.descriptors[key].name}</span>
            <button className="DescriptorList-item-delete" onClick={() => this.props.deleteDescriptor(key)}>-</button>
          </li>
        ))}
      </ul>
    );
  }
}

export default DescriptorList;
