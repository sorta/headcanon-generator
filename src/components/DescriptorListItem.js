import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Descriptor extends Component {
  static propTypes = {
    descriptor: PropTypes.shape({
      name: PropTypes.string,
    }),
    descriptorAvailability: PropTypes.object,
    updateDescriptor: PropTypes.func,
    deleteDescriptor: PropTypes.func,
    setAvailability: PropTypes.func,
    isManaging: PropTypes.func,
    index: PropTypes.string,
  };
  static defaultProps = {
    descriptor: { name: '' },
    descriptorAvailability: {},
    updateDescriptor: () => {},
    deleteDescriptor: () => {},
    setAvailability: () => {},
    isManaging: () => false,
    index: '',
  };

  handleChange = (event) => {
    const target = event.currentTarget;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const updatedDescriptor = {
      ...this.props.descriptor,
      [target.name]: value,
    };

    this.props.updateDescriptor(this.props.index, updatedDescriptor);
  };

  // setAvailability = (typeKey, itemKey, value)
  toggleAvailability = (event) => {
    // console.log(event);
    this.props.setAvailability('descriptor', this.props.index, event.currentTarget.checked);
  };

  render() {
    if (!this.props.isManaging()) {
      return (
        <li key={this.props.index} className="DescriptorListItem list-item control-row">
          <span className="DescriptorList-item-text">
            {this.props.descriptor.name}
          </span>
        </li>
      );
    }

    return (
      <li key={this.props.index} className="DescriptorListItem list-item control-row">
        <input
          name="isAvailable"
          type="checkbox"
          checked={!({}.hasOwnProperty.call(this.props.descriptorAvailability, this.props.index))}
          onChange={this.toggleAvailability}
        />
        <input
          name="name"
          type="text"
          placeholder="Descriptor Name"
          className="control-stretch"
          value={this.props.descriptor.name}
          onChange={this.handleChange}
        />
        <button
          className="DescriptorListItem-delete"
          onClick={() => this.props.deleteDescriptor(this.props.index)}
        >-</button>
      </li>
    );
  }
}

export default Descriptor;
