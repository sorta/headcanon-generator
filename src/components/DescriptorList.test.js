import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import DescriptorList from './DescriptorList';
import DescriptorListItem from './DescriptorListItem';

const props = {
  descriptors: {
    d_test_1: { name: 'Test 1' },
    d_test_2: { name: 'Test 2' },
  },
  descriptorAvailability: {},
  updateDescriptor: jest.fn(),
  deleteDescriptor: jest.fn(),
  setAvailability: jest.fn(),
  isManaging: () => false,
};

// Tests for regular mode
describe("DescriptorList", () => {

  it('renders and matches our snapshot', () => {
    const snapshot = renderer.create(
      <DescriptorList { ...props } />
    );
    const tree = snapshot.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has a DescriptorList ul element', () => {
    const component = shallow(<DescriptorList { ...props } />);
    expect(component.find('ul.DescriptorList.options-list').exists()).toBe(true);
  });

  it('has right number of DescriptorListItems', () => {
    const component = shallow(<DescriptorList { ...props } />);
    expect(component.find(DescriptorListItem)).toHaveLength(Object.keys(props.descriptors).length);
  });
});
