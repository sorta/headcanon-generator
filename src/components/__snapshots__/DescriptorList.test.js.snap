// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescriptorList renders and matches our snapshot 1`] = `
<ul
  className="options-list DescriptorList"
>
  <li
    className="DescriptorListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_desc_d_test_1"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="DescriptorList-item-text"
      htmlFor="isAvailable_desc_d_test_1"
    >
      Test 1
    </label>
  </li>
  <li
    className="DescriptorListItem list-item control-row"
  >
    <input
      checked={true}
      className="cb-fake"
      id="isAvailable_desc_d_test_2"
      name="isAvailable"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="DescriptorList-item-text"
      htmlFor="isAvailable_desc_d_test_2"
    >
      Test 2
    </label>
  </li>
</ul>
`;
